src/sftp.rs

```diff
@@ -371,11 +393,36 @@ impl Sftp {
             Err(Error::from_errno(rc))
         }
     }
+
+    fn unwrap_raw_or_err(&self) -> Result<*mut raw::LIBSSH2_SFTP, Error> {
+        match self.inner.as_ref() {
+            Some(inner) => Ok(inner.raw),
+            None => Err(Error::from_errno(raw::LIBSSH2_ERROR_BAD_USE)),
+        }
+    }
+
+    #[doc(hidden)]
+    pub fn shutdown(&mut self) -> Result<(), Error> {
+        let raw = self.unwrap_raw_or_err()?;
+        self.rc(unsafe { raw::libssh2_sftp_shutdown(raw) })?;
+        let _ = self.inner.take();
+        Ok(())
+    }
 }
 
 impl Drop for Sftp {
     fn drop(&mut self) {
-        unsafe { assert_eq!(raw::libssh2_sftp_shutdown(self.raw), 0) }
+        // Set ssh2 to blocking if sftp was not shutdown yet.
+        if let Some(inner) = self.inner.as_ref() {
+            let was_blocking = inner.sess.is_blocking();
+            if !was_blocking {
+                inner.sess.set_blocking(true);
+            }
+            assert_eq!(unsafe { raw::libssh2_sftp_shutdown(inner.raw) }, 0);
+            if !was_blocking {
```

Comment:
I'd probably eliminate the conditionals; I don't think there's any material cost associated with them:

```rust
let was_blocking = inner.sess.is_blocking();
inner.sess.set_blocking(true);
assert_eq!(unsafe { raw::libssh2_sftp_shutdown(inner.raw) }, 0);
inner.sess.set_blocking(was_blocking);
```